2019-03-Autumn/2_string_manipulation/README.md

```diff
@@ -86,20 +107,159 @@ TODO :bug:
+2. Use a suffix tree.
+
+We will discuss the memoization strategy; investigating suffix trees 
+is left as an exercise to interested students.
+
+##### What is "Memoization"?
+
+Memoization is a strategy from **dynamic programming**.
+
+When we solve a small subset of a problem, we would like to have easy 
+access to this solution so that we don't have to resolve the problem 
+all over again later. This goes to the heart of **memoization**: we create 
+a mechanism to reference the same solution later.
+
+##### Using Memoization in This Solution
+
+Our memoization strategy is to record the number of consecutive 
+preceding characters for each possible comparison between `s1` and `s2`. 
+
+Because this comparison must be made for each combination of indeces in `s1` 
+and `s2`, we will use a 2-dimensional array to store these values. 
+
+When we can reference this array which we will call `found`, we can 
+perform the following steps:
+
+1. Loop over `s1` and `s2` to get all possible index combinations (`i` and `j`)
+2. For each index pair:
+    a. If either `i` or `j` is `0`, we can't possibly check preceding characters, so we enter `0` in `found[i][j]`
+    b. Otherwise, if `s1[i] == s2[j]`, we have found part of a common substring.
+    c. If we have a common substring, we check the preceding character pair `s1[i - 1]` and `s2[j - 1]`; if these are equal, the substring extends to previous characters. We take the value in `found[i - 1][j - 1]`, add 1 to it, and store it in `found[i][j]`.
+    d. If `s1[i] != s2[j]`, we store `0` in `found[i][j]`.
+    e. Last but not least, we want to update our `result` value if `s1[i] == s2[j]` and our previous `result` value represents a shorter string. We can easily use the value in `found[i][j]` to determine the substring's length, the indeces of the substring's location, or the substring itself. The implementation of this is dependent on what you chose to return from the function; the example provided here returns the index bounds of the substring in `s1` so we can produce any of the 3 interesting outputs.
+
+A Python implementation of this is as follows:
+
+```python3
+# an implementation with memoization
+def longest_substring_dynamic(s1: str, s2: str, found: list) -> tuple:
+    indeces = (-1, -1)
+    for i in range(len(s1)):
+        for j in range(len(s2)):
+            # base case: beginning of string, can't check previous chars
+            if i == 0 or j == 0:
+                found[i][j] = 0
+            # general case: check for identical neighbors in each string
+            elif s1[i] == s2[j] and s1[i - 1] == s2[j - 1]:
+                # increment if characters are consecutive
+                found[i][j] = found[i - 1][j - 1] + 1
+            # failure case: consecutive char count resets
+            else:
+                found[i][j] = 0
+            indeces = (i - found[i][j], i + 1) if found[i][j] + 1 >= indeces[1] - indeces[0] and s1[i] == s2[j] else indeces
```

Comment:
```suggestion
            # update indeces
            if s1[i] == s2[j] and found[i][j] + 1 >= indeces[1] - indeces[0]:
                indeces = (i - found[i][j], i + 1)
```

Reply:
Um, i added a commit for this? See 4867a16d4f4beb69746ec4a65e610166d271599b??

Reply:
Oops, forgot about the solutions file